examples/refit_freesolv.py

```diff
@@ -0,0 +1,830 @@
+        futures.append(client.submit(loss_serial, x, serial_loss_args))
+    results = [fut.result() for fut in futures]
+
+    # Combine results [((loss, aux), g), ...]
+    v = np.sum([result[0][0] for result in results])
+    aux = {}
+    for result in results:
+        aux.update(result[0][1])
+    grad = np.sum([result[1] for result in results], axis=0)
+    return (v, aux), grad
+
+
+def loss(x, loss_args: LossArgs) -> Tuple[float, Gradient]:
+    """
+    Main loss function. This will:
+        1) compute the loss using the new parameters
+        2) determine if new samples should be generated
+        3) generate new samples and update the prefactors
+        4) recompute the loss using the new samples
+
+    Parameters
+    ----------
+    x:
+        Current compressed charge parameters.
+    loss_args:
+        Contains auxiliary information needed to compute the predicted dG.
+
+    Returns
+    -------
+    loss:
+        Value for the loss.
+    grad:
+        Gradient with resepect to x.
+    """
+    # compute the loss
+    (v, aux), g = loss_parallel(loss_args.n_gpus, x, loss_args)
+
+    mols = loss_args.mols
+    resampled_mols = mols_to_resample(mols, aux)
+    if len(resampled_mols) > 0:
+        print("Resampling", len(resampled_mols), "molecules")
+
+    # update ff params
+    _, expand = get_compress_expand_fxns(loss_args.initial_q_params, loss_args.active_q_types)
+    new_ff = copy.deepcopy(loss_args.ff)
+    new_ff.q_handle.params = expand(x)
+
+    if len(resampled_mols):
+        # generate new samples if needed
+        samples = generate_samples_for_mols_parallel(loss_args.n_gpus, resampled_mols, loss_args.prefactor_dict, new_ff)
+        samples = [samples[get_mol_name(mol)] for mol in resampled_mols]
+
+        # recompute the prefactors for the new samples
+        ref_delta_fs = [None] * len(resampled_mols)  # Will be updated below
+        new_prefactor_dict = compute_prefactors(expand(x), new_ff, resampled_mols, samples, ref_delta_fs)
+
+        # update ref_delta_f to match the new samples
+        update_delta_fs(resampled_mols, loss_args.prefactor_dict, new_prefactor_dict, new_ff)
+
+        # overwrite the old prefactors
+        loss_args.prefactor_dict.update(new_prefactor_dict)
+
+        # recompute the loss
+        (v, aux), g = loss_parallel(loss_args.n_gpus, x, loss_args)
+
+        # check that no molecules are marked to be resampled
+        new_resampled_mols = mols_to_resample(resampled_mols, aux)
+        assert (
+            len(new_resampled_mols) == 0
+        ), f"no molecules should need to be resampled here {[get_mol_name(mol) for mol in new_resampled_mols]}"
+
+    # update the predictions to pass back to main
+    pred_dg_dict = {mol_name: a["pred_dg"] for mol_name, a in aux.items()}
+    loss_args.pred_dg_dict.update(pred_dg_dict)
+
+    # Normalize loss/grads by total number of molecules
+    return float(v / len(mols)), np.array(g / len(mols), dtype=np.float64)
```

Comment:
Why is this normalization necessary?

```suggestion
    return float(v), np.array(g, dtype=np.float64)
```

Reply:
It's to make it easier to interpret the loss independent of the number of molecules.

Reply:
I think it would be preferable to move the normalization outside of this function.

(So that
(1) `loss` has consistent behavior with `_loss_serial`, `loss_serial`, `loss_parallel`, which all return `sum` instead of `mean`,
(2) `loss(..., dataset_a) + loss(..., dataset_b) == loss(..., combine(dataset_a, dataset_b))`
(3) we don't risk accidentally dividing by `len(mols)` twice
(4) we won't need to remember to multiply by `len(mols)` if we want to add parameter regularization (`regularize(params) + loss(params, data)`, where `loss` is expected to be size-extensive with `len(data)`.)